content/advent-2019/safe-use-of-unsafe-pointer.md

```diff
@@ -0,0 +1,322 @@
+./main.go:20:33: possible misuse of unsafe.Pointer
+```
+
+In order to understand this error, we must consult [the rules](https://golang.org/pkg/unsafe/#Pointer)
+of the `unsafe.Pointer` type:
+
+> Converting a Pointer to a uintptr produces the memory address of the value
+> pointed at, as an integer. The usual use for such a uintptr is to print it.
+>
+> Conversion of a uintptr back to Pointer is not valid in general.
+>
+> A uintptr is an integer, not a reference. Converting a Pointer to a uintptr
+> creates an integer value with no pointer semantics. Even if a uintptr holds
+> the address of some object, the garbage collector will not update that
+> uintptr's value if the object moves, nor will that uintptr keep the object
+> from being reclaimed.
+
+We can isolate the offending portion of our program as follows:
+
+```go
+p := uintptr(unsafe.Pointer(&arr[0]))
+
+// What happens if there's a garbage collection here? It's possible that the
+// address (now stored as a uintptr integer) is no longer valid!
+//
+// Let's assume p no longer points at a uint32. Perhaps this memory is now being
+// used to store a user's authentication credentials or a TLS private key.
+//
+// We've introduced a potential security flaw in our application!
+fmt.Printf("%d ", (*(*uint32)(unsafe.Pointer(p))))
+```
+
+In effect, once we've converted an `unsafe.Pointer` to `uintptr`, we cannot
+safely convert it back to `unsafe.Pointer`, with the exception of one special
+case:
+
+> If p points into an allocated object, it can be advanced through the object by
+> conversion to uintptr, addition of an offset, and conversion back to Pointer.
+
+In order to perform this pointer arithmetic iteration logic safely, we must
+perform the type conversions and pointer arithmetic all at once:
+
+```go
+package main
+
+import (
+    "fmt"
+    "unsafe"
+)
+
+func main() {
+    // An array of contiguous uint32 values stored in memory.
+    arr := []uint32{1, 2, 3}
+
+    // The number of bytes each uint32 occupies: 4.
+    const size = unsafe.Sizeof(uint32(0))
+
+    for i := 0; i < len(arr); i++ {
+        // Print an integer to the screen by:
+        //   - taking the address of the first element of the array
+        //   - applying an offset of (i * 4) bytes to advance into the array
+        //   - converting the uintptr back to *uint32 and dereferencing it to
+        //     print the value
+        fmt.Printf("%d ", *(*uint32)(unsafe.Pointer(
+            uintptr(unsafe.Pointer(&arr[0])) + (uintptr(i) * size),
+        )))
+    }
+}
+```
+
+This program produces the same result as before, but is now considered valid by
+`go vet` as well!
+
+```
+$ go run main.go 
+1 2 3 
+$ go vet .
+```
+
+I don't recommend using pointer arithmetic for iteration logic in this way, but
+it's excellent that Go provides this escape hatch (and tooling for using it
+safely!) when it is truly needed.
+
+## Run-time verification with the Go compiler's `-d=checkptr` flag
```

Comment:
You have to use `-gcflags=all=-d=checkptr` at this moment to make `-d=checkptr` works correctly. The reason is that without `all=`, the `-d=checkptr` is only applied for your current package. If your package uses function from other package, which is mark as `go:nocheckptr`, but that package is not compiled with `-d=checkptr`, then it will make false positive. `reflect.Value.Pointer` and `reflect.Value.UnsafeAddr` are examples.

Reply:
Updated!

Reply:
For upcoming Go release, you don't have to use `all=-d=checkptr` anymore. See https://go-review.googlesource.com/c/go/+/222878